### PR TITLE
Improve GitLab projects name verification

### PR DESCRIPTION
**Pull request by DarkaMaul**
This PR addresses https://github.com/pypi/warehouse/issues/15852

Notably, it prevents names from : 
- being in the reserved GitLab name list
- ending with `.atom` or `.git`
- ending with something else than `a-zA-Z0-9`
- consecutive special characters

The keywords list are extracted from [GitLab code](https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/gitlab/path_regex.rb#L16). As noted by @facutuesca , this is mostly a cosmetic change, so having unsynchronized lists should not introduce a security risk.

Of note, we could slightly improve the regexes by limiting the number of potential matches to a fixed upper bound and prevent ReDos types of attacks.

### CODE REVIEW

File: warehouse/oidc/forms/gitlab.py [hunk 1]
```diff
@@ -19,11 +19,117 @@
...
+  "sitemap",
+  "sitemap.xml",
+  "sitemap.xml.gz",
+  "slash-command-logo.png",
+  "snippets",
+  "unsubscribes",
+  "uploads",
+  "users",
+  "v2",
+}
```
Comment (woodruffw) [inline warehouse/oidc/forms/gitlab.py:92]: I'm tempted to say that we shouldn't bother with these lists -- they're not going to be stable and they're almost certainly already incomplete/will go out of sync with GitLab over time.

**Comment by woodruffw** [comment]
Thanks @DarkaMaul! I want to pare this back a bit to avoid depending on assumptions internal to GitLab's code, but the consecutive character check makes a lot of sense to me to add.

File: warehouse/oidc/forms/gitlab.py [hunk 2]
```diff
@@ -19,11 +19,117 @@
...
+        for char in field.data:
+            match char:
+                case char if char in self.forbidden_characters:
+                    countdown -= 1
+
+                    if countdown == 0:
+                        raise wtforms.validators.ValidationError(self.message)
 
+                case _:
+                    countdown = 2
```
[the remark this answers is not in the document]
Comment (DarkaMaul) [reply warehouse/oidc/forms/gitlab.py]: Almost - I used an inverted one because the pattern should match.

`(?!.*[._-]{2})`

(And indeed, that's a much better solution than the previous one)

**Comment by di** [comment]
Is this ready for review?

**Comment by DarkaMaul** [comment]
I think we are still unsure if we want to introduce a dependency on Gitlab internal code (for the list of forbidden project names).

I can refactor the PR to only include the following rules : 
- no consecutive special characters
- no ending with a special character

We can leave the two others out of this PR:
- not a reserved name
- does not ends with `.git` or `.atom`

**Comment by facutuesca** [comment]
IMO having those lists of forbidden names is fine, since they are documented on [GitLab's docs](https://docs.gitlab.com/ee/user/reserved_names.html#reserved-project-names).  For me the important question is if having such exhaustive checking is worth it. @di what do you think?

**Comment by di** [comment]
The purpose of these validators is to prevent likely typos or misunderstandings about what the value of this field should be.

I think it's fairly unlikely that a user would frequently be trying to use these forbidden names, and even if they were, it would be fairly clear why it was failing.

TL;DR: I don't think being so exhaustive is necessary or worth it.

**Comment by di** [comment]
> does not ends with .git or .atom

I think this could happen frequently enough that it's worth including.

**Comment by woodruffw** [comment]
> > does not ends with .git or .atom
> 
> I think this could happen frequently enough that it's worth including.

Agreed! Let's cover those two suffixes but leave the reserved names out.

**Review by woodruffw — APPROVED** [review]
LGTM, thank you @DarkaMaul!

**Review by firefly9966 — APPROVED** [review]

**Review by di — APPROVED** [review]
